packages/react/src/__tests__/ReactElementValidator-test.js

```diff
@@ -248,61 +214,48 @@ describe('ReactElementValidator', () => {
-      );
-      expect(console.error.calls.argsFor(1)[0]).toBe(
-        'Warning: React.createElement: type is invalid -- expected a string ' +
-          '(for built-in components) or a class/function (for composite ' +
-          'components) but got: null.',
-      );
-      expect(console.error.calls.argsFor(2)[0]).toBe(
-        'Warning: React.createElement: type is invalid -- expected a string ' +
-          '(for built-in components) or a class/function (for composite ' +
-          'components) but got: boolean.',
-      );
-      expect(console.error.calls.argsFor(3)[0]).toBe(
-        'Warning: React.createElement: type is invalid -- expected a string ' +
-          '(for built-in components) or a class/function (for composite ' +
-          'components) but got: object.',
-      );
-      expect(console.error.calls.argsFor(4)[0]).toBe(
-        'Warning: React.createElement: type is invalid -- expected a string ' +
-          '(for built-in components) or a class/function (for composite ' +
-          'components) but got: object. You likely forgot to export your ' +
-          "component from the file it's defined in, or you might have mixed up " +
-          'default and named imports.',
-      );
-    }
-    React.createElement('div');
-    if (__DEV__) {
-      expect(console.error.calls.count()).toBe(5);
-    }
+    expect(() => {
+      React.createElement(undefined);
+      React.createElement(null);
+      React.createElement(true);
+      React.createElement({x: 17});
+      React.createElement({});
+    }).toWarnDev([
+      'Warning: React.createElement: type is invalid -- expected a string ' +
+        '(for built-in components) or a class/function (for composite ' +
+        'components) but got: undefined. You likely forgot to export your ' +
+        "component from the file it's defined in, or you might have mixed up " +
+        'default and named imports.',
+      'Warning: React.createElement: type is invalid -- expected a string ' +
+        '(for built-in components) or a class/function (for composite ' +
+        'components) but got: null.',
+      'Warning: React.createElement: type is invalid -- expected a string ' +
+        '(for built-in components) or a class/function (for composite ' +
+        'components) but got: boolean.',
+      'Warning: React.createElement: type is invalid -- expected a string ' +
+        '(for built-in components) or a class/function (for composite ' +
+        'components) but got: object.',
+      'Warning: React.createElement: type is invalid -- expected a string ' +
+        '(for built-in components) or a class/function (for composite ' +
+        'components) but got: object. You likely forgot to export your ' +
+        "component from the file it's defined in, or you might have mixed up " +
+        'default and named imports.',
+    ]);
+
+    // Should not log any additional warnings
+    expect(() => React.createElement('div')).toWarnDev([]);
```

Comment:
Is this necessary? Shouldn’t an unexpected log fail the test anyway? I expect that if we explicitly mark each warning block then any warning outside should be a failure. Maybe even immediately when it’s called rather than at the end of the test.

Reply:
> Shouldn’t an unexpected log fail the test anyway

Good point.

Reply:
Now that I look at this again, I believe this explicit check is necessary since we've necessarily spied on `console.error` in order for the matcher to work.

IIR, I originally tried to clear the spy at the end of the matcher, so unexpected errors/warnings would cause an error, but neither the version of Jest we're using (`21.3.0-beta.4`) nor the latest release (22) actually expose the `spy.mockReset` method [the docs mention](http://facebook.github.io/jest/docs/en/jest-object.html#jestspyonobject-methodname).

We could call `spy.and.callThrough()` at the bounds of the matcher, so that errors were actually passed through, but (a) that could negatively affect tests with multiple `.toWarnDev` calls and (b) this wouldn't actually fail the test, only print the red text to the console.

Maybe I'm overlooking something?

Reply:
>Now that I look at this again, I believe this explicit check is necessary since we've necessarily spied on console.error in order for the matcher to work.

Wouldn't this approach solve it? https://github.com/facebook/react/pull/11786#discussion_r155672766

>IIR, I originally tried to clear the spy at the end of the matcher, so unexpected errors/warnings would cause an error

I don't quite understand what clearing means in this context, or why it would be helpful.

My proposal would be to set up a mock when we enter `toWarn` and tear it down when we exit it. Outside of `toWarn`, I'd want `console.error` to be replaced with a function that insta-throws.

Reply:
I think we're saying the same thing, but misunderstanding each other.

My initial design for this matcher was to setup a spy/mock when entered and then remove it before existing. This way unexpected `console.error` calls would fail the test (as they do if you don't spy in the first place). The Jest docs [suggest this should be possible](http://facebook.github.io/jest/docs/en/jest-object.html#jestspyonobject-methodname) via a `mockReset` method, but that method doesn't actually exist (in Jest 21 or 22) so I'm not sure why the docs mention it. The problem then, is that once we've spied on `console.error` initially, we can't remove that spy and so any unexpected calls to it don't actually cause failures.

We could work around this in other ways. For example, I could not use Jest's `spyOn` functionality at all- but just override `console.error` with my own function and then restore the originally one before exiting. Maybe that's best anyway?

Reply:
>We could work around this in other ways. For example, I could not use Jest's spyOn functionality at all- but just override console.error with my own function and then restore the originally one before exiting. Maybe that's best anyway?

Sorry I wasn't clear; yes, this is what I meant.

Reply:
>My initial design for this matcher was to setup a spy/mock when entering and then remove it before exiting. This way unexpected console.error calls would fail the test (as they do if you don't spy in the first place)

They still wouldn't fail it *immediately*. Only upon checking the mock (which currently happens at the end of the test). But this is annoying because you have to "backtrack" to see *where exactly* an extra warning has fired and why.

My proposal was to override these methods so we can *replace them with shims that throw* when we're outside of the `toWarn` block. This way calling them would fail the test *immediately* rather than we have reached the end of the test. But it does require us to roll our own.

Reply:
I think that's actually a reasonable plan for this matcher, since we don't need full the spy functionality as implemented by Jasmine/Jest anyway.

Reply:
Oh sorry, looks like our comments got crossed.

WRT just rolling our own "spy"- cool. I'll do this.

Regarding the second thing you say, I think you're just suggesting we modify our `console` overrides [here](https://github.com/facebook/react/blob/247c5241706356033cd5bf1dc31197e058853e13/scripts/jest/setupTests.js#L58-L87) to throw immediately rather than in an `afterEach`? I don't think this is specifically related to this PR, but I'd be happy to roll it into this one. 😄

Reply:
>Regarding the second thing you say, I think you're just suggesting we modify our console overrides here to throw immediately rather than in an afterEach? I don't think this is specifically related to this PR, but I'd be happy to make that change as well.

It's related to making this PR more than an aesthetic change :-) I think throwing early extracts a benefit from the new API—something we haven't been able to do before. It also prevents accidental use of old API (directly asserting on the mock). But I agree it's not a prerequisite.

Reply:
Hmm, ok. Although I think aesthetics/ergonomics are important for test writing.

> I think throwing early extracts a benefit from the new API—something we haven't been able to do before.

Sorry if I'm being dense, but I don't understand what you mean about us being unable to throw immediately before. Could you explain?

We always wanted unexpected warnings to fail, and we always signified they were expected by spying on them. That hasn't changed with this PR.

Reply:
Never mind. I believe I understand now.

Reply:
With this API:

```js
spyOnDev(console, 'warn'); // line 1

console.warn('good warning'); // line 2
console.warn('unexpected warning'); // line 3
doSomethingElse(); // line 4

expectDev(console.warn.calls.length).toBe(1); // line 5
```

if there is a second (unexpected) warning, the test fails on line 5. You have to hunt `console.error` calls to see that the unexpected warning was caused by line 3 (in practice, it's somewhere inside the library).

With this API:

```js
expect(() => { // line 1
  console.warn('good warning'); // line 2
  console.warn('unexpected warning'); // line 3
}).toWarnInDev(['stuff']) // line 4

console.warn('stuff'); // line 5
doSomethingElse(); // line 6
```

the test fails directly on line 5, with a callstack pointing to the `console.warn` call. We don't need to wait for the whole section to execute (or to check the expected warning count twice) to know it has failed. And we have the exact stack trace for *where* the unexpected warning happened.

-----

Similarly, another scenario before and after:

```js
doSomething();
console.warn('did not expect this warning');
doSomethingElse();
```

In our current setup this test fails *after* it has finished executing. We have no clue as to *where in the code* the warning has happened.

Compared to the same code:

```js
doSomething();
console.warn('did not expect this warning');
doSomethingElse();
```

Since we now know that any warning outside of `toWarn` is bad, we can have it throw immediately, and thus show the stack trace.

---

If this still doesn't make sense, never mind, maybe I'm not explaining it well and it's not that important anyway. All I'm saying is that `toWarn()` kind of API gives us the ability to know where exactly the "expected zone" for warnings starts and ends, as opposed to the previous API whose granularity was "something in this test warns somewhere". So previously we couldn't fail immediately on a warning outside of that zone since the whole test was "fair game". But now we know exactly which lines can warn, and if any line outside warns, we can fail.

In fact now we can validate the warning messages immediately.

With

```js
expect(() => {
  console.error('unexpected stuff');
  doSomething();
  console.error('stuff');
  console.error('other stuff');
}).toWarn([
  'stuff',
  'other stuff'
]);
```

we can now throw right in the first `console.error` call because *we already know it has mismatched* and there's no point executing the test further (and obscuring what happened by letting variables be mutated, etc).

On the other hand, with the current approach

```js
spyOnDev(console, 'error');

console.error('unexpected stuff');
doSomething();
console.error('stuff');
console.error('other stuff');

expect(console.errors.calls.length).toBe(2);
```

we only know something went wrong by the time we reached the last line, and the state might have been obscured by now. Usually I end up having to comment out every line one by one until I find the earliest one where the mismatched warning still happens.

Reply:
(My initial comment didn't make sense so I edited it)

Reply:
Ah! I think I understand what you're saying better now.

The boundary is clearer with the new API. Before the boundary was more informal, between `spyOn` and the first `(...calls.length).toBe()` check.

It was still possible to do the throw-immediate thing before though. That's what confused me a bit.

Reply:
I guess it was also easier for the old spies to miss things if warnings were logged after the last count check too. That's another advantage of this boundary.

Reply:
>It was still possible to do the throw-immediate thing before though

How would it work with this test case?

```js
spyOnDev(console, 'error')

console.error('unexpected warning')
console.error('expected warning')

expect(console.error.calls.count).toBe(1)
expect(console.error.calls.argsFor(0)[0]).toBe('expected warning')
```

Previously we couldn't safely throw on the first `console.error` because we didn't know its message is unexpected by the time it ran.

But with `toWarn` we can because we know the exhaustive list of expected messages (and their count) *before* `console.error` is called.

Reply:
It just occurred to me coming back from lunch that, since I know the error message(s) ahead of time, I could also throw immediately (within my fake spy) any time a message is logged that isn't expected.

I'll rework the matcher to behave this way instead.